README.md
# Tip 基础提示框

---
[![Build Status](https://img.shields.io/travis/aralejs/dialog.svg?style=flat)](https://travis-ci.org/aralejs/dialog)
[![Coverage Status](https://img.shields.io/coveralls/aralejs/dialog.svg?style=flat)](https://coveralls.io/r/aralejs/dialog)

提示框组件，专注于带箭头的气泡框显示、隐藏、动画等交互功能。
模块内部集成了模板和样式，无须单独引用样式，可以很方便地直接使用。

> 注：1.2.0 版本后 Tip 模块被原来的子模块 Atip 模块所替换。

---

Tip 模块继承自 [popup](http://aralejs.org/popup/)，可使用包括 [overlay](http://aralejs.org/overlay/)、[widget](http://aralejs.org/widget/)、[base](http://aralejs.org/base/)、[class](http://aralejs.org/class/)、[events](http://aralejs.org/events/)、[attribute](http://aralejs.org/base/docs/attribute.html)、[aspect](http://aralejs.org/base/docs/aspect.html) 的属性和方法。

## 配置说明

### trigger `selector`

触发点元素。

### triggerType `string`

触发方式，[`hover`|`click`|`focus`]。

### content `string`

提示内容，支持字符和 html 结构。

### arrowPosition `number`

箭头位置，按钟表点位置，目前支持1、2、5、7、10、11点位置，默认 7 点钟。

![图示](https://i.alipayobjects.com/e/201307/jBty06lQT.png)

### theme `string`

配色主题，内置黄色、蓝色和白色三种提示框样式 [`yellow`|`blue`|`white`]，默认为 `yellow`。

### inViewport `boolean`

当弹出层显示在屏幕外时，是否自动转换浮层位置，默认为 false。

### align `object`

参照 Overlay 的 [align](http://aralejs.org/overlay/#align-object) 配置，
指定此定位参数时，认为开发者要自定义提示框位置，此时 `arrowPosition` 属性失效。

---

## 最佳实践

```js
import Tip from 'arale-tip';
new Tip({
    trigger: '#test',       // 触发器
    content: '我是内容',    // 提示框显示的内容
    arrowPosition: 11       // 11点钟的箭头位置
});
```

其他用法请参见 examples


HISTORY.md
# 历史记录

---
## 2.0.0
`changed` spm2npm 

## 1.3.0

`tag:changed` 迁移到 spm@3.x

## 1.2.2

`tag:improved` 升级 popup 到 1.1.6

## 1.2.1

`tag:improved` 升级 tip 依赖的默认样式 alice/poptip 到 [1.2.0](http://aliceui.org/poptip/history.html)。

## 1.2.0

`tag:changed` #8 去掉 tip 模块，用 atip 替换了 tip 模块。

`tag:changed` 现在可以指定 align 参数用来自定义位置，此时 arrowPosition 失效。

`tag:improved` 升级 popup 到 1.1.5


## 1.1.4

`tag:improved` 升级 popup 到 1.1.2

`tag:improved` 升级 atip 依赖的默认样式 alice/poptip 到 1.1.1 ，相应修改了默认模板。


## 1.1.3

`tag:fixed` 由于之前 build 出来的代码包含依赖两个不同版本的 position, 重新 build

`tag:improved` 升级 popup 到 1.1.1

`tag:improved` 添加 styleBox, 解决样式冲突问题。


## 1.1.2

`tag:improved` 依赖升级到 [arale.popup@1.1.0](http://aralejs.org/popup/history.html)。

`tag:fixed` [pull#5](https://github.com/aralejs/tip/pull/6) 修复当有多个trigger时，定位总是指向第一个trigger的问题。（相关代码已经在popup中实现了）


## 1.1.1

`tag:improved` 依赖升级到 [arale.popup@1.0.2](http://aralejs.org/popup/history.html)。

`tag:unresolved` 当有多个trigger时，定位总是指向第一个trigger。


## 1.1.0

`tag:improved` 依赖升级到 [arale.popup@1.0.1](http://aralejs.org/popup/history.html)。

`tag:improved` 添加了 `pointPos` 参数，Atip 也可以使用此参数进行位置调节。[#5](https://github.com/aralejs/tip/issues/5)

## 1.0.0

`tag:improved` 依赖升级到 [arale.popup@1.0.0](http://aralejs.org/popup/history.html)。

`tag:improved` Atip 依赖了 [alice.poptip](http://aliceui.org/poptip) ，Atip 组件无须再单独引用样式。


## 0.9.4

`tag:improved` 依赖升级到 [arale.popup@0.9.12](http://aralejs.org/popup/history.html)。


## 0.9.3

`tag:improved` 依赖升级到 [arale.popup@0.9.10](http://aralejs.org/popup/history.html)，从而支持简单动画效果，以及修复了一些 bug。


## 0.9.2

`tag:new` 将 alipay.atip 移到本组件中。


tests/tip-spec.js
var exepct = require('expect.js');

var $ = require('jquery');
var Tip = require('../index');
var tip, trigger;
var CONTENT = '_content_';

describe('Tip', function () {

  beforeEach(function () {
    trigger = $('<div style="width:200px;">trigger</div>').appendTo(document.body);
  });

  afterEach(function () {
    tip.element.remove();
    trigger.remove();
  });

  it('content', function () {
    tip = new Tip({
      content: CONTENT
    });
    tip.show();
    expect(tip.element.find('[data-role=content]').html()).to.be(CONTENT);
  });

  it('hover trigger', function (done) {
    tip = new Tip({
      trigger: trigger,
      content: CONTENT
    });
    expect(tip.element.is(':visible')).to.be(false);
    trigger.trigger('mouseover');
    setTimeout(function () {
      expect(tip.element.is(':visible')).to.be(true);
      done();
    }, 80);
  });

  it('content from function', function () {
    tip = new Tip({
      trigger: trigger,
      content: function () {
        return 'test content';
      }
    });
    tip.show();
    expect(tip.element.find('[data-role="content"]').html()).to.be('test content');
  });

  it('stylesheet load', function () {
    tip = new Tip({
      trigger: trigger,
      content: CONTENT
    });
    tip.show();
    var color = tip.element.css('color');
    expect(color == 'rgb(219, 124, 34)' || color == '#db7c22').to.be(true);
  });

  it('set size', function () {
    tip = new Tip({
      trigger: trigger,
      content: CONTENT,
      width: 600,
      height: '1024px'
    });
    tip.show();
    expect(tip.element.find('[data-role="content"]').width()).to.be(600);
    expect(tip.element.find('[data-role="content"]').height()).to.be(1024);
  });

  it('themes', function () {
    tip = new Tip({
      trigger: trigger,
      content: CONTENT
    });
    tip.show();
    expect(tip.get('theme')).to.be('yellow');
    expect(tip.element.hasClass('ui-poptip-yellow')).to.be(true);
    tip.set('theme', 'white');
    expect(tip.element.hasClass('ui-poptip-yellow')).to.be(false);
    expect(tip.element.hasClass('ui-poptip-white')).to.be(true);
    tip.set('theme', 'blue');
    expect(tip.element.hasClass('ui-poptip-yellow')).to.be(false);
    expect(tip.element.hasClass('ui-poptip-white')).to.be(false);
    expect(tip.element.hasClass('ui-poptip-blue')).to.be(true);
  });

  it('arrowPosition', function () {
    tip = new Tip({
      trigger: trigger,
      content: CONTENT
    });
    tip.show();
    expect(tip.get('arrowPosition')).to.be(7);
    expect(tip.get('direction')).to.be('up');
    tip.set('arrowPosition', 10);
    expect(tip.get('direction')).to.be('right');
    tip.set('arrowPosition', 2);
    expect(tip.get('direction')).to.be('left');
    tip.set('arrowPosition', 1);
    expect(tip.get('direction')).to.be('down');
    tip.set('arrowPosition', 5);
    expect(tip.get('direction')).to.be('up');
  });

  it('inViewport top', function () {
    trigger.css({
      position: 'absolute',
      top: 0,
      left: 10
    });
    tip = new Tip({
      trigger: trigger,
      content: CONTENT,
      arrowPosition: 7,
      height: 100,
      inViewport: true
    });
    expect(tip.get('arrowPosition')).to.be(7);
    tip.show();
    expect(tip.get('arrowPosition')).to.be(11);
    tip.hide();
    trigger.css({
      top: 'auto',
      bottom: 0
    });
    tip.show();
    expect(tip.get('arrowPosition')).to.be(7);
    trigger.css({
      position: null,
      top: null,
      left: null
    });
  });

  it('inViewport bottom', function () {
    trigger.css({
      position: 'absolute',
      bottom: 0,
      left: 10
    });
    tip = new Tip({
      trigger: trigger,
      content: CONTENT,
      arrowPosition: 1,
      height: 100,
      inViewport: true
    });
    expect(tip.get('arrowPosition')).to.be(1);
    tip.show();
    expect(tip.get('arrowPosition')).to.be(5);
    tip.hide();
    trigger.css({
      top: 0,
      bottom: 'auto'
    });
    tip.show();
    expect(tip.get('arrowPosition')).to.be(1);
    trigger.css({
      position: null,
      top: null,
      left: null
    });
  });

  it('ignore arrowPosition when there is align', function () {
    tip = new Tip({
      trigger: trigger,
      arrowPosition: 11,
      content: CONTENT,
      align: {
        baseXY: [0, 0],
        selfXY: [0, 0]
      }
    });
    tip.show();
    expect(tip.element.offset().top).to.be(trigger.offset().top);
    expect(tip.element.offset().left).to.be(trigger.offset().left);
  });

});

src/atip.js
define(function(require, exports, module) {

    var $ = require('$');
    var Tip = require('./tip');

    // 依赖样式 alice.poptip    
    require('ui-poptip');

    // 气泡提示弹出组件
    // ---
    var Atip = Tip.extend({

        attrs: {

            template: require('./atip.tpl'),

            // 提示内容
            content: '这是一个提示框',

            // 箭头位置
            // 按钟表点位置，目前支持1、2、5、7、10、11点位置
            arrowPosition: 7,

            // 颜色 [yellow|blue|white]
            theme: 'yellow',

            // 当弹出层显示在屏幕外时，是否自动转换浮层位置
            inViewport: false,

            // 宽度
            width: 'auto',

            // 高度
            height: 'auto'

        },

        setup: function() {
            this._originArrowPosition = this.get('arrowPosition');
            // 设置位置，避免滚动条闪动
            this.element.css({
                position: "absolute",
                top: 0,
                left: 0
            });
            Atip.superclass.setup.call(this);
        },

        show: function() {
            Atip.superclass.show.call(this);
            this._makesureInViewport();
        },

        _makesureInViewport: function() {
            if (this.get('inViewport')) {
                var ap = this._originArrowPosition,
                    scrollTop = $(window).scrollTop(),
                    viewportHeight = $(window).outerHeight(),
                    elemHeight = this.element.height() + this.get('distance'),
                    triggerTop = $(this.get('trigger')).offset().top,
                    arrowMap = {
                        '1': '5',
                        '5': '1',
                        '7': '11',
                        '11': '7'
                    };
                if ((ap === 11 || ap === 1) &&
                    (triggerTop > scrollTop + viewportHeight - elemHeight)) {
                    this.set('arrowPosition', arrowMap[ap]);
                }
                else if ((ap === 7 || ap === 5) &&
                          (triggerTop < scrollTop + elemHeight)) {
                    this.set('arrowPosition', arrowMap[ap]);
                }
                else {
                    this.set('arrowPosition', this._originArrowPosition);
                }
            }
        },

        // 用于 set 属性后的界面更新

        _onRenderArrowPosition: function(val, prev) {
            val = parseInt(val, 10);
            var arrow = this.$('.ui-poptip-arrow');
            arrow.removeClass('ui-poptip-arrow-' + prev)
                 .addClass('ui-poptip-arrow-' + val);

            var direction = '', arrowShift = 0;
            if (val === 10) {
                direction = 'right';
                arrowShift = 20;
            }
            else if (val === 11) {
                direction = 'down';
                arrowShift = 22;
            }
            else if (val === 1) {
                direction = 'down';
                arrowShift = -22;
            }
            else if (val === 2) {
                direction = 'left';
                arrowShift = 20;
            }
            else if (val === 5) {
                direction = 'up';
                arrowShift = -22;
            }
            else if (val === 7) {
                direction = 'up';
                arrowShift = 22;
            }
            this.set('direction', direction);
            this.set('arrowShift', arrowShift);
            this._setAlign();
        },

        _onRenderWidth: function(val) {
            this.$('[data-role="content"]').css('width', val);
        },

        _onRenderHeight: function(val) {
            this.$('[data-role="content"]').css('height', val);
        },

        _onRenderTheme: function(val, prev) {
            this.element.removeClass('ui-poptip-' + prev);
            this.element.addClass('ui-poptip-' + val);
        }

    });

    module.exports = Atip;

});




lib/basic-tip.js
'use strict';

var _aralePopup = require('arale-popup');

var _aralePopup2 = _interopRequireDefault(_aralePopup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 通用提示组件
// 兼容站内各类样式
module.exports = _aralePopup2.default.extend({

  attrs: {
    // 提示内容
    content: null,

    // 提示框在目标的位置方向 [up|down|left|right]
    direction: 'up',

    // 提示框离目标距离(px)
    distance: 8,

    // 箭头偏移位置(px)，负数表示箭头位置从最右边或最下边开始算
    arrowShift: 22,

    // 箭头指向 trigger 的水平或垂直的位置
    pointPos: '50%'
  },

  _setAlign: function _setAlign() {
    var alignObject = {},
        arrowShift = this.get('arrowShift'),
        distance = this.get('distance'),
        pointPos = this.get('pointPos'),
        direction = this.get('direction');

    if (arrowShift < 0) {
      arrowShift = '100%' + arrowShift;
    }

    if (direction === 'up') {
      alignObject.baseXY = [pointPos, 0];
      alignObject.selfXY = [arrowShift, '100%+' + distance];
    } else if (direction === 'down') {
      alignObject.baseXY = [pointPos, '100%+' + distance];
      alignObject.selfXY = [arrowShift, 0];
    } else if (direction === 'left') {
      alignObject.baseXY = [0, pointPos];
      alignObject.selfXY = ['100%+' + distance, arrowShift];
    } else if (direction === 'right') {
      alignObject.baseXY = ['100%+' + distance, pointPos];
      alignObject.selfXY = [0, arrowShift];
    }

    alignObject.comeFromArrowPosition = true;
    this.set('align', alignObject);
  },

  // 用于 set 属性后的界面更新
  _onRenderContent: function _onRenderContent(val) {
    var ctn = this.$('[data-role="content"]');
    if (typeof val !== 'string') {
      val = val.call(this);
    }
    ctn && ctn.html(val);
  }

});

lib/tip.js
'use strict';

var _jquery = require('jquery');

var _jquery2 = _interopRequireDefault(_jquery);

var _basicTip = require('./basic-tip');

var _basicTip2 = _interopRequireDefault(_basicTip);

require('alice-poptip');

require('./tip.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 气泡提示弹出组件
// ---
var Tip = _basicTip2.default.extend({
  attrs: {
    template: '<div class="ui-poptip">\n                  <div class="ui-poptip-shadow">\n                  <div class="ui-poptip-container">\n                      <div class="ui-poptip-arrow">\n                          <em></em>\n                          <span></span>\n                      </div>\n                      <div class="ui-poptip-content" data-role="content">\n                      </div>\n                  </div>\n                  </div>\n              </div>',

    // 提示内容
    content: 'A TIP BOX',

    // 箭头位置
    // 按钟表点位置，目前支持1、2、5、7、10、11点位置
    // https://i.alipayobjects.com/e/201307/jBty06lQT.png
    arrowPosition: 7,

    align: {
      setter: function setter(val) {
        // 用户初始化时主动设置了 align
        // 且并非来自 arrowPosition 的设置
        if (val && !val.comeFromArrowPosition) {
          this._specifiedAlign = true;
        }
        return val;
      }
    },

    // 颜色 [yellow|blue|white]
    theme: 'yellow',

    // 当弹出层显示在屏幕外时，是否自动转换浮层位置
    inViewport: false
  },

  setup: function setup() {
    _basicTip2.default.superclass.setup.call(this);
    this._originArrowPosition = this.get('arrowPosition');

    this.after('show', function () {
      this._makesureInViewport();
    });
  },

  _makesureInViewport: function _makesureInViewport() {
    if (!this.get('inViewport')) {
      return;
    }
    var ap = this._originArrowPosition,
        scrollTop = (0, _jquery2.default)(window).scrollTop(),
        viewportHeight = (0, _jquery2.default)(window).outerHeight(),
        elemHeight = this.element.height() + this.get('distance'),
        triggerTop = this.get('trigger').offset().top,
        triggerHeight = this.get('trigger').height(),
        arrowMap = {
      '1': 5,
      '5': 1,
      '7': 11,
      '11': 7
    };

    if ((ap == 11 || ap == 1) && triggerTop + triggerHeight > scrollTop + viewportHeight - elemHeight) {
      // tip 溢出屏幕下方
      this.set('arrowPosition', arrowMap[ap]);
    } else if ((ap == 7 || ap == 5) && triggerTop < scrollTop + elemHeight) {
      // tip 溢出屏幕上方
      this.set('arrowPosition', arrowMap[ap]);
    } else {
      // 复原
      this.set('arrowPosition', this._originArrowPosition);
    }
  },

  // 用于 set 属性后的界面更新
  _onRenderArrowPosition: function _onRenderArrowPosition(val, prev) {
    val = parseInt(val, 10);
    var arrow = this.$('.ui-poptip-arrow');
    arrow.removeClass('ui-poptip-arrow-' + prev).addClass('ui-poptip-arrow-' + val);

    // 用户设置了 align
    // 则直接使用 align 表示的位置信息，忽略 arrowPosition
    if (this._specifiedAlign) {
      return;
    }

    var direction = '',
        arrowShift = 0;
    if (val === 10) {
      direction = 'right';
      arrowShift = 20;
    } else if (val === 11) {
      direction = 'down';
      arrowShift = 22;
    } else if (val === 1) {
      direction = 'down';
      arrowShift = -22;
    } else if (val === 2) {
      direction = 'left';
      arrowShift = 20;
    } else if (val === 5) {
      direction = 'up';
      arrowShift = -22;
    } else if (val === 7) {
      direction = 'up';
      arrowShift = 22;
    }
    this.set('direction', direction);
    this.set('arrowShift', arrowShift);
    this._setAlign();
  },

  _onRenderWidth: function _onRenderWidth(val) {
    this.$('[data-role="content"]').css('width', val);
  },

  _onRenderHeight: function _onRenderHeight(val) {
    this.$('[data-role="content"]').css('height', val);
  },

  _onRenderTheme: function _onRenderTheme(val, prev) {
    this.element.removeClass('ui-poptip-' + prev);
    this.element.addClass('ui-poptip-' + val);
  }

});

module.exports = Tip;